### PR TITLE
post: instead-of-dto

### PR DESCRIPTION
**Pull request by g4s8**
New post about DTOs

### CODE REVIEW

**Comment by g4s8** [comment]
@Evgeniy-jin this is a new post, please take a look.

File: _posts/2019-09-06-instead-of-dto.md [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+<a href="https://en.wikipedia.org/wiki/XML" target="_blank">XML</a>
+to pass data from mail to service,
+it has some advantages over previos solutions:
+ - validation - we can enforce the protocol with `xsd` schemas and fail method `accept()` if xml is invalid
+ - queries - `MailService` can use xpath queries to access the data
+ - readability - XML has readable format, so it's easier to view XML file instead of using debugger to inspect DTO instances
+ - allows you to build complex data structures - XML is a flexible language to define complex data structures
+ - transformations - XML data structure can be transformed using `xsl` transformations
+ - flexability - mail object can construct data message from internal state by configured logic in `send` method or decorate
+ actor (`Mail`) or receiver `MailService` by decorators to put additional data.
```
Comment (Evgeniy-jin) [inline _posts/2019-09-06-instead-of-dto.md]: 1) an actor
2) a receiver
3) к чему относится 'by decorators'?

File: _posts/2019-09-06-instead-of-dto.md [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+
+The disatvantages are:
+ - complexity - it's too complex for simple data structures
+ - knowledge - it requires additional knowledge in XML language to design such structures
+
+
+It's better to start with `xsd` schema to define data structure, but
+it'll be over-complex for simple blog post, so I skip it here.
```
Comment (Evgeniy-jin) [inline _posts/2019-09-06-instead-of-dto.md]: for a simple blog post

File: _posts/2019-09-06-instead-of-dto.md [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+If you are not famialiar with `xsd` schemas you may start learning it
```
Comment (Evgeniy-jin) [inline _posts/2019-09-06-instead-of-dto.md]: learning them

File: _posts/2019-09-06-instead-of-dto.md [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+        throws IOException {
+        final String address = mail.xpath("/mail/recipient/text()").get(0);
+        final String subject = mail.xpath("/mail/subject/text()").get(0);
+        final List<String> ccs = mail.xpath("/mail/ccs/cc");
+        // TODO: send via SMTP
+    }
+}
+```
+
+On the other side, `Mail` implementations can use
```
Comment (Evgeniy-jin) [inline _posts/2019-09-06-instead-of-dto.md]: On the other side - с другой стороны? Тогда, on the other hand - это устойчивое выражение

File: _posts/2019-09-06-instead-of-dto.md [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+<a href="https://github.com/yegor256/xembly" target="_blank">Xembly</a>
+language to build `XML` object message using directives
+(pay attention: this class doesn't expose internals and doesn't break encapsulation, it's rather
```
Comment (Evgeniy-jin) [inline _posts/2019-09-06-instead-of-dto.md]: it rather

File: _posts/2019-09-06-instead-of-dto.md [hunk 6]
```diff
@@ -0,0 +1,278 @@
...
+constructing a message to another object using internal state):
```
Comment (Evgeniy-jin) [inline _posts/2019-09-06-instead-of-dto.md]: constructs a message

File: _posts/2019-09-06-instead-of-dto.md [hunk 7]
```diff
@@ -0,0 +1,278 @@
...
+from encapsulated state and pass it to `accept()` of `MailService`:
+```java
+new MailSimple(
+  "Test mail",
+  "Hello",
+  "test@test.com"
+).post(new Smpt(connection));
+```
+
+One of the advantages is flexability:
```
Comment (Evgeniy-jin) [inline _posts/2019-09-06-instead-of-dto.md]: flexibility

File: _posts/2019-09-06-instead-of-dto.md [hunk 8]
```diff
@@ -0,0 +1,278 @@
...
+new MailWithCC(
+  new MailSimple(
+    "Test mail",
+    "Hello",
+    "test@test.com"
+  ),
+  "copy@test.com"
+).post(new Smpt(connection));
+```
+Or service itself can be decorated in similar way:
```
Comment (Evgeniy-jin) [inline _posts/2019-09-06-instead-of-dto.md]: in a similar way:

File: _posts/2019-09-06-instead-of-dto.md [hunk 9]
```diff
@@ -0,0 +1,278 @@
...
+            ).value()
+          )
+      )
+    )
+  }
+}
+```
+
+Another important advantage of this approach is that it's easy to unit-test
+this classes:
```
Comment (Evgeniy-jin) [inline _posts/2019-09-06-instead-of-dto.md]: these classes

File: _posts/2019-09-06-instead-of-dto.md [hunk 10]
```diff
@@ -0,0 +1,278 @@
...
+    mail -> MatcherAssert.assertThat(
+      mail.node(),
+      XhtmlMatchers.hasXPaths("/mail/ccs/cc[./text() = 'copy@test.com']")
+    )
+  );
+}
+```
+
+When you change your data format (and update `xsd` schema), you may write
+`xsl` transformation to update old version to new on "data side", so you just change
```
Comment (Evgeniy-jin) [inline _posts/2019-09-06-instead-of-dto.md]: 1. an `xsl` transformation
2. the old version
3. to the new one on 'data side'

File: _posts/2019-09-06-instead-of-dto.md [hunk 11]
```diff
@@ -0,0 +1,278 @@
...
+the code to support only new format and applying transformations to convert old data.
```
Comment (Evgeniy-jin) [inline _posts/2019-09-06-instead-of-dto.md]: 1. the new format
2. and apply
3. the old data

File: _posts/2019-09-06-instead-of-dto.md [hunk 12]
```diff
@@ -0,0 +1,278 @@
...
+
+
+## Conclusion
+
+To summarize it all - we're spending more time on implementation, since we need
+to write all these schemas, xml manipulators, etc, but keeping much more time
```
Comment (Evgeniy-jin) [inline _posts/2019-09-06-instead-of-dto.md]: but saving (?) much more time

File: _posts/2019-09-06-instead-of-dto.md [hunk 13]
```diff
@@ -0,0 +1,278 @@
...
+on maintaining the code and making it more readable. But you always should think
+about the balance between the cost of implementing and cost of maintaining:
+I'd never use XML language for simple data messages, e.g. if by business requirements
+mail can contain onle message and address, nothing more; it would be easier to
```
Comment (Evgeniy-jin) [inline _posts/2019-09-06-instead-of-dto.md]: 1. a mail
2. only 
3. a message and address

File: _posts/2019-09-06-instead-of-dto.md [hunk 14]
```diff
@@ -0,0 +1,278 @@
...
+put these properties right in the method, since creating XML definitions for that case will
+be too expensive for project.
```
Comment (Evgeniy-jin) [inline _posts/2019-09-06-instead-of-dto.md]: for the project

File: _posts/2019-09-06-instead-of-dto.md [hunk 15]
```diff
@@ -0,0 +1,278 @@
+---
+layout: post
+title: Instead of DTO
+---
+
+Objects are using data messages to communicate with each other,
+it means that object methods can accept some data, but data structure
+sometimes is too complex. When complex data message is designed wrongly
+it tends to low maintainability because it becomes harder to test this code,
```
[the remark this answers is not in the document]
Comment (g4s8) [reply _posts/2019-09-06-instead-of-dto.md]: @Evgeniy-jin can I say "it tends to reduce maintainability" (make it lower)?

[the remark this answers is not in the document]
Comment (Evgeniy-jin) [reply _posts/2019-09-06-instead-of-dto.md]: I like that

**Pull request merged**